Enable Firewall on Node, Add Windows Firewall rules for required ports

raghushantha opened the pull request:
**What this PR does / why we need it**:
Secures K8's Windows Nodes by enabling Firewall.
Allows only Master node to connect to kubelet ports
Creates Firewall exceptions for container websockets for communicating to PowerShell Console workloads

The fixes have been verified in DF cluster.

**Release note**:

```release-note
- Enable Firewall (with exceptions) for Windows Nodes to lockdown access to the nodes
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/azure/acs-engine/725)
<!-- Reviewable:end -->

raghushantha committed:
Enable Firewall on Node, Add Windows Firewall rules for required ports

robbiezhang commented:
@yangl900

robbiezhang commented on parts/kuberneteswindowssetup.ps1:
```diff
@@ -201,8 +201,17 @@ Get-PodGateway(`$podCIDR)
+    netsh advfirewall firewall add rule name="Container: Allow access to node localport 8080" dir=in action=allow protocol=TCP localport=8080
+    netsh advfirewall firewall add rule name="Container: Allow access to node localport 8888" dir=in action=allow protocol=TCP localport=8888
+    netsh advfirewall firewall add rule name="Container: Allow UDP inbound traffic for Container DNS Port 53" dir=in action=allow localport=53 protocol=UDP
+    netsh advfirewall firewall add rule name="Node: Allow only K8 Master to access localport 4194" dir=in action=allow protocol=TCP localport=4194 remoteip=`${global:MasterIP}
```
is the global:MasterIP internal IP or external?

robbiezhang replied:
I believe their communication are through internal IPs. In this case, you need to think about supporting multiple master nodes, although cloud shell only use 1 master node now.

raghushantha replied:
@robbiezhang To test the firewall rules, I used the Master's well known private IP [10.240.255.5], since this is a well known acs-engine constant.
So yes, this is internal.

Looking at the script [line 190], the global variable is set to the script parameter value.

What other changes need to happen to support multiple master nodes? I don't see any reference to this in the codebase

robbiezhang replied:
Multi-master is an ACS-Engine feature. It will introduce a LoadBalancer for the master nodes. The IP address for the LB is 10.240.255.15. However, there is no such LB in the single master node cluster. So the outbound IP address is the node address (10.240.255.5). Do you know how this script handle this?

raghushantha replied:
Multi-Master must be handled in this script. Anthony's team owns this, I believe.

For now, the windows script is being called from the engine's go template:
https://github.com/Azure/acs-engine/blob/d3059c436d30bdc196d76cda27b1f051719316e7/pkg/acsengine/engine.go#L663

JiangtianLi replied:
The script is executed as part of VM extensions and the parameters passed to it are in kuberneteswinagentresourcesvmas.t. MasterIP is variables('kubernetesAPIServerIP')

JiangtianLi replied:
@colemickens Could you please take a look at port allowed?

JiangtianLi commented on parts/kuberneteswindowssetup.ps1:
```diff
@@ -201,8 +201,17 @@ Get-PodGateway(`$podCIDR)
 {
-    # Turn off Firewall to enable pods to talk to service endpoints. (Kubelet should eventually do this)
-    netsh advfirewall set allprofiles state off
+    # Windows Firewall rules to allow only Master to access Node's kubelet ports
```
Are these rules only for powershell console workloads and this PR is not going to be merged into master branch, right?

raghushantha replied:
@JiangtianLi The rules are not just for PSCloudShell - but there are no other workloads on the windows nodes. We want to lockdown the Windows nodes by allowing only master to communicate with Kubelet ports and allowing PSCloudShell websocket connections.

Additional windows customers can add their own exceptions as they are onboarded.

We want the firewall feature to be in PROD. Will the release to PROD happen from Master OR Migration branch?

JiangtianLi commented on parts/kuberneteswindowssetup.ps1:
```diff
@@ -201,8 +201,17 @@ Get-PodGateway(`$podCIDR)
+    netsh advfirewall firewall add rule name="Node: Allow only K8 Master to access localport 10255" dir=in action=allow protocol=TCP localport=10255 remoteip=`${global:MasterIP}
+    
+    # Turn-on the firewall since we have allowed access to required ports
+    netsh advfirewall set allprofiles state on
```
We explicitly turn off firewall for allprofiles in Set-DockerNetwork. I think it should be ON already but I can't guarantee. Anyway, it doesn't hurt to turn it on.

raghushantha replied:
Correct, after adding firewall rules it is better to turn on the firewall for all profiles

JiangtianLi commented on parts/kuberneteswindowssetup.ps1:
```diff
@@ -201,8 +201,17 @@ Get-PodGateway(`$podCIDR)
+    netsh advfirewall firewall add rule name="Container: Allow access to node localport 8080" dir=in action=allow protocol=TCP localport=8080
+    netsh advfirewall firewall add rule name="Container: Allow access to node localport 8888" dir=in action=allow protocol=TCP localport=8888
+    netsh advfirewall firewall add rule name="Container: Allow UDP inbound traffic for Container DNS Port 53" dir=in action=allow localport=53 protocol=UDP
+    netsh advfirewall firewall add rule name="Node: Allow only K8 Master to access localport 4194" dir=in action=allow protocol=TCP localport=4194 remoteip=`${global:MasterIP}
```
@raghushantha  Could you add comment for the port allowed? How do the rules allow web service on port 80 or other service on custom port?

raghushantha replied:
Will add comments in the script to explain each rule. I thought the rule names are self-explanatory -:).

We only allow websocket connections to be made to PSCloudShell (8080/8888). No other customer's ports are allowed. Also, since this is a first party service, any other customer using the Windows kube nodes need to bring in their own rules. This change will lock down the nodes for PSCloudShell.

JiangtianLi replied:
Thanks. Specifically please comment what each port is used for. (53 is obvious though).

raghushantha committed:
Added comments for firewall rules

raghushantha committed:
Allow all traffic; lockdown kubectl Node ports to Master only

raghushantha committed:
Remove & and single quote in comment section

resource group deployment parser does not like these chars in the comment section

raghushantha commented:
Fix has been verified with a DF cluster deployment. fyi

JiangtianLi commented:
@raghushantha Thanks. Have you gone through sanity test with k8s windows example workload?

raghushantha commented:
@JiangtianLi We verified using PSCloudShell workload. This is similar to https://github.com/Azure/acs-engine/blob/master/docs/kubernetes.windows.md.

Basically we ensured that traffic flows from webbrowser to the frontend and then to the container websockets.

JiangtianLi commented:
@raghushantha Please verify k8s windows example workload since they use different port - just in case.

raghushantha commented:
@JiangtianLi Verified simpleweb and hybrid scenarios as outlined in https://github.com/Azure/acs-engine/blob/master/docs/kubernetes.windows.md

yangl900 commented:
looks good to me!
